File: Circle.cpp

#include <Circle.h>
#include <iostream>

/*!
    \file
    \brief The circle class methods.

    This file contains the code that implements the Circle structure methods.
*/

CircFit::Circle::Circle(float x, float y, float radius)
{
    h = x;
    k = y;
    r = radius;
    if (r>0)
        undefined = false;
    else
        undefined = true;
}

CircFit::Circle::Circle(std::vector<cv::Point> &pts, std::vector<bool> &inliers)
{
    fitCircle(pts, inliers);
}

float CircFit::Circle::fitCircle(std::vector<cv::Point> &pts, std::vector<bool> &inliers, float w, float sigma, float p)
{
    unsigned int nInliers;
    
    h = k = r = 0;
    undefined = true;
    if (pts.size() <3)
        return -1;
    if (pts.size() == 3)
    {
        fitCircle(pts[0].x, pts[0].y, pts[1].x, pts[1].y, pts[2].x, pts[2].y);
        if (!undefined)
            return 0;
        else
            return -1;
    }
    return ransacFit(pts, inliers, nInliers, w, sigma, p);
}


void CircFit::Circle::fitCircle( float x1, float y1, float x2, float y2, float x3, float y3)
{
    float den;

    den = x2*y3+x1*(y2-y3)-x3*y2+y1*(x3-x2);

    if (den != 0.0)
    {
        float dsq1, dsq2, dsq3;

        dsq1 = -y1*y1-x1*x1;
        dsq2 = -y2*y2-x2*x2;
        dsq3 = -y3*y3-x3*x3;

        den = 1.0/den;

        h = -0.5*den*((y2-y3)*dsq1+(y3-y1)*dsq2+(y1-y2)*dsq3);
        k = -0.5*den*((x3-x2)*dsq1+(x1-x3)*dsq2+(x2-x1)*dsq3);
        r = sqrt(h*h+k*k-den*((x2*y3-x3*y2)*dsq1+(x3*y1-x1*y3)*dsq2+(x1*y2-x2*y1)*dsq3));
        undefined = false;
    }
    else
    {
        h = k = r = 0;
        undefined = true;
    }
}

float CircFit::Circle::fitBestCircle(std::vector<cv::Point> &pts, unsigned int nInliers, unsigned int *idx)
{
    unsigned int i;
    float error = 0.;

    if (nInliers < 3)
    {
        h = k = r = 0;
        undefined = true;
        return -1.;
    }

    if (nInliers == 3)
    {
        float x[3], y[3];
        for (i=0;i<3 ; ++i)
        {
            x[i] =pts[idx[i]].x;
            y[i] =pts[idx[i]].y;
        }
        fitCircle( x[0], y[0], x[1], y[1], x[2], y[2]);
        if (!undefined)
            return 0.;
        else
            return -1.;
    }
    else
    {
        cv::Mat A, b, x;
        
        A = cv::Mat::zeros(cv::Size(3, nInliers), CV_32FC1); 
        b = cv::Mat::zeros(cv::Size(1, nInliers), CV_32FC1);
        for (i=0;i<nInliers;++i)
        {
            A.at<float>(i,0) = pts[idx[i]].x;
            A.at<float>(i,1) = pts[idx[i]].y;
            A.at<float>(i,2) = 1.;
            b.at<float>(i, 0) = -pow(pts[idx[i]].x, 2)-pow(pts[idx[i]].y, 2);
        }
        if (!cv::solve(A, b, x, cv::DECOMP_SVD))
        {
            std::cerr << "Singluar Matrix at fitCircle(std::vector<cv::Point> &pts..." << std::endl << std::endl;
            h = k = r = 0;
            undefined = true;
        }
        h = -0.5 * x.at<float>(0,0);
        k = -0.5 * x.at<float>(1,0);
        r = sqrt(h*h+k*k-x.at<float>(2,0));
        undefined = false;
        for (i=0;i<nInliers;++i)
            error += pow(distMin(pts[idx[i]]),2);
        error /= nInliers;
        
        return error;
    }
}

void CircFit::Circle::selectAndTestSample(std::vector <cv::Point> &pts, std::vector<bool> &inliers, float thr, unsigned int &nInliers, unsigned int &nOutliers)
{
    unsigned int l, inl;
    unsigned int idx1, idx2, idx3;
    float d;
    std::vector<cv::Point>::iterator it,end;
    std::vector<bool>::iterator itInl;

    l = pts.size();
    if (l >= 3)
    {
      if (inliers.size() != pts.size())
         inliers = std::vector<bool>(l, false);
    
        do
        {
            idx1 = (int)(l * drand48());
            idx2 = (int)(l * drand48());
            idx3 = (int)(l * drand48());
        } while (idx3 == idx1 || idx3 == idx2 || idx1 == idx2);

        undefined = true;
        h = k = r = 0;
        fitCircle( pts[idx1].x, pts[idx1].y, pts[idx2].x, pts[idx2].y, pts[idx3].x, pts[idx3].y);
        it = pts.begin();
        end = pts.end();
        itInl = inliers.begin();
        inl = 0;

        d = distMin(*it);
        if (d <= thr)
        {
            *itInl = true;
            inl++;
        }
         else
            *itInl = false;
        

        //Deal with the rest but the last one.
        for(++it, ++itInl; it+1 != end; it++, itInl++)
        {
            d = distMin(*it);
            if (d <= thr)
            {
               *itInl = true;
               inl++;
            }
            else
               *itInl = false;
        }
        //Deal with the last one.
         d = distMin(*it);
         if (d <= thr)
         {
            *itInl = true;
            inl++;
         }
         else
            *itInl = false;
        nInliers = inl;
        nOutliers = l - inl;
    }
}

// \var p  Es la probabilidad de que un inlier sea una inliers. (p=0.99)
// \var w  Es la proporcion de inliers vs outliers.
float CircFit::Circle::ransacFit(std::vector <cv::Point> &pts, std::vector<bool> &inliers, unsigned int &nInl, float w, float sigma, float p)
{
    float error, thr = 3.84* sigma;
    unsigned int i, j, l, N, T, nInliers, nOutliers, best;
    unsigned int *bestInl, *itBest;
    std::vector<cv::Point>::iterator it, end;
    std::vector<bool>::iterator itInl;

    srand48(time(0));

    l = pts.size();
    if (l<3)
    {
        nInl=0;
        undefined = true;
        return -1;
    }
    if (inliers.size() != l)
      inliers = std::vector<bool>(l, false);

    N = log(1.-p)/log(1-pow(w,3)); //El numero máximo de iteraciones que vamos a realizar.
    T=(unsigned int)(l * w); //El valor de inliers esperado.
    best = 0;
    bestInl = new unsigned int[l];
    for (i=0;i<N;++i)
    {
        selectAndTestSample(pts, inliers, thr, nInliers, nOutliers);
        if (nInliers > best)
        {
            best = nInliers;
            it = pts.begin();
            end = pts.end();
            itInl = inliers.begin();
            itBest = bestInl;
            for (j = 0;it != end; ++j, ++it, ++itInl)
                if (*itInl)
                {
                    *itBest = j; 
                    itBest++;
                }
        }
        if (nInliers>T)
            break;
    }
    error = fitBestCircle(pts, best, bestInl);
    for (i=0;i<pts.size();++i)
        inliers[i] = false;
    for (i=0;i<best;++i)
        inliers[bestInl[i]] = true;

    delete[] bestInl;
    nInl = best;
    return error;
}

float CircFit::Circle::distMin(cv::Point &p)
{
    return fabs(sqrt(pow(h-p.x,2)+pow(k-p.y,2))-r);
}

File: Circle.py
#! /usr/bin/env python3
#encoding: utf-8

import numpy as np
import numpy.random as rand
import time

class Circle:
   def __init__(self):
      self.h = self.k = self.r = 0.
      self.undefined = True

   def setCircle(self, x, y, radius):
      self.h, self.k, self.r = x, y, radius
      if self.r > 0:
         self.undefined = False
      else:
         self.h = self.k = self.r = 0.
         self.undefined = True

   def fitCircle(self, pts, w=0.6, sigma=1.0, p=0.99):
      self.h = self.k = self.r = 0.
      self.undefined = True
      if len(pts) < 3:
         return
      if len(pts) == 3:
         self.fitCircle3P(pts)
         if not self.undefined:
            return 0
      else:
         self.ransacFit(pts, nInliers, w, sigma, p)
         return nInliers

   def fitCircle3P(self, pts):
      x1, y1 = pts[0, 0], pts[0, 1]
      x2, y2 = pts[1, 0], pts[1, 1]
      x3, y3 = pts[2, 0], pts[2, 1]

      den = x2 * y3 + x1 * (y2 - y3) - x3 * y2 + y1 * (x3 - x2)

      if den != 0.0:
         dsq1 = -y1 * y1 - x1 * x1
         dsq2 = -y2 * y2 - x2 * x2
         dsq3 = -y3 * y3 - x3 * x3

         den = 1.0/den;

         self.h = -0.5*den*((y2-y3)*dsq1+(y3-y1)*dsq2+(y1-y2)*dsq3);
         self.k = -0.5*den*((x3-x2)*dsq1+(x1-x3)*dsq2+(x2-x1)*dsq3);
         self.r = np.sqrt(self.h**2+self.k**2-den*((x2*y3-x3*y2)*dsq1+(x3*y1-x1*y3)*dsq2+(x1*y2-x2*y1)*dsq3));
         self.undefined = False;
      else:
         self.h = self.k = self.r = 0.
         self.undefined = True;

   def fitBestCircle(self, pts):
      error = 0.
      N = len(pts)
      if N < 3:
         self.h = self.k = self.r = 0.
         self.undefined = True
         return -1

      if N == 3 :
         self.fitCircle3P(pts)
         return 0

      pr, pc = pts.shape
      if pc < 3:
         pts = np.concatenate([pts, np.ones((pr, 1))], 1)
      else:
         pts[:, 2] = np.ones(pr)
      A = pts
      b = (-pts[:,0]**2-pts[:,1]**2).reshape((len(pts),1))

      X = np.linalg.lstsq(A, b, rcond=None)
      x = X[0]
      self.h = -0.5 * x[0];
      self.k = -0.5 * x[1];
      self.r = np.sqrt(self.h ** 2 + self.k ** 2 - x[2])
      self.undefined = True
      error = 0.
      for i in range(N):
         error = error + self.distMin(pts[i]) ** 2;
      error = error / N
      return error

   def selectAndTestSample(self, pts, thr):
      pr,pc = pts.shape
      if pc < 3:
         return -1
      N = len(pts)
      if N >= 3:
         idx1 = rand.randint(0, N)
         idx2 = rand.randint(0, N)
         idx3 = rand.randint(0, N)
         while idx3 == idx1 or idx3 == idx2 or idx1 == idx2:
            idx1 = rand.randint(0, N)
            idx2 = rand.randint(0, N)
            idx3 = rand.randint(0, N)
         self.undefined = False 
         self.h = self.k = self.r = 0.
         self.fitCircle3P(pts[[idx1, idx2, idx3], :])
         inl = 0
         idx=[]
         for i in range(N):
            d = self.distMin(pts[i])
            if d <= thr:
               idx.append(i)
         return idx

   def ransacFit(self, pts, w, sigma =1 , p = 0.99):
      thr = 3.84 * sigma
      rand.seed(int(time.time()))

      l = len(pts)
      if l < 3:
         self.undefined = True
         self.h = self.k = self.r = 0.
         return 0
      pr,pc = pts.shape
      if pr < 3 or pc < 2:
         self.undefined = True
         return -1, -1
      if pc < 3:
         pts = np.concatenate([pts, np.zeros((pr,1))])
      N = int(np.ceil(np.log(1. - p)/np.log(1.-w**3)))
      print("N = ", N)
      T = int(l * w)
      best = 0
      for i in range(N):
         inliers = self.selectAndTestSample(pts, thr)

         nInliers = len(inliers)
         nOutliers = N - nInliers 
         if nInliers > best:
            best = nInliers
            bestInl = np.array(inliers)
            if best != len(bestInl):
               print("GROSSO ERRORE", best, len(bestInl))
         if nInliers > T:
            break
      pts[:,2] = np.zeros(l)
      pts[bestInl, 2] = 1
      error = self.fitBestCircle(pts[bestInl,:])
      return best, error
          
   def distMin(self, p):
      return np.fabs(np.sqrt((self.h-p[0])**2+(self.k-p[1])**2)-self.r)

